includes/API/ContextAuthorized.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace DataAccounting\API;
+
+use MediaWiki\Permissions\PermissionManager;
+use MediaWiki\Rest\HttpException;
+use MediaWiki\Rest\SimpleHandler;
+use LogicException;
+use RequestContext;
+
+abstract class ContextAuthorized extends SimpleHandler {
+
+	/**
+	 * @var PermissionManager 
+	 */
+	protected $permissionManager;
+
+	/**
+	 * @param PermissionManager $permissionManager
+	 */
+	public function __construct( PermissionManager $permissionManager ) {
+		$this->permissionManager = $permissionManager;
+	}
+
+	public function execute() {
+		$paramSettings = $this->getParamSettings();
+		$validatedParams = $this->getValidatedParams();
+		$unvalidatedParams = [];
+		$params = [];
+		foreach ( $this->getRequest()->getPathParams() as $name => $value ) {
+			$source = $paramSettings[$name][self::PARAM_SOURCE] ?? 'unknown';
+			if ( $source !== 'path' ) {
+				$unvalidatedParams[] = $name;
+				$params[] = $value;
+			} else {
+				$params[] = $validatedParams[$name];
+			}
+		}
+
+		if ( $unvalidatedParams ) {
+			throw new LogicException(
+				'Path parameters were not validated: ' . implode( ', ', $unvalidatedParams )
+			);
+		}
+
+		$this->checkPermission( ...$params );
+
+		// @phan-suppress-next-line PhanUndeclaredMethod
+		// @phpstan-ignore-next-line
+		return $this->run( ...$params );
+	}
+
+	/** @inheritDoc */
+	public function needsWriteAccess() {
+		return false;
+	}
+
+	/** @inheritDoc */
+	public function needsReadAccess() {
+		return false;
+	}
+
+	/**
+	 * @return string[]
+	 */
+	protected function getPermissions(): array {
+		return [ 'read' ];
+	}
+
+	protected function checkPermission() {
+		// @phan-suppress-next-line PhanUndeclaredMethod
+		// @phpstan-ignore-next-line
+		$title = $this->provideTitle( ...func_get_args() );
+		if ( !$title ) {
+			throw new HttpException( "No title provided for permission check", 403 );
+		}
+		$user = RequestContext::getMain()->getUser();
+		foreach ( $this->getPermissions() as $permission ) {
+			if ( $this->permissionManager->userCan( $permission, $user, $title ) ) {
+				continue;
+			}
+			throw new HttpException(
+				"Permission denied for '$permission' on '{$title->getFullText()}'"
```

Comment:
@pwirth is this always the error message regardless of whether the page with that title exists in the DB? Just wondering of the plausible deniability of the page.

Reply:
the message thrown, when no title was provided by extending class is actually:
"No title provided for permission check"
or the message thrown by extending class when there was an invalid rev_id, a page could not be calculated from
I added these scenarios to the tests

Reply:
I think we should throw 404 not found for all those cases, so that an attacker can't tell which page title is present in the PKC. Do you agree with this, @FantasticoFox ?

Reply:
Agreed.

Reply:
Additionally, if it is all 403, then an attacker can tell that it is a private wiki.

Reply:
i changed all the permission related errors to a basic 404 "Not found"